Tasks/Npm/npmtask.ts

```diff
@@ -1,27 +1,194 @@
+        };
+
+        var code : number = await npmAuthRunner.exec(execOptions);
+        tl.debug('Auth helper exitted with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        // ignore any auth failure and try to run the task.
+        tl.debug(tl.loc('NpmAuthFailed', err.message));
+    }
+}
+
+async function tryRunNpmConfigAsync(npmConfigRunner: trm.ToolRunner, execOptions : trm.IExecOptions) {
+    try {
+        var code = await npmConfigRunner.exec(execOptions);
+    } catch (err) {
+        // do not throw on this failure.
+        tl.warning(tl.loc('NpmConfigFailed', err.Message));
+    }
+}
+
+async function runNpmCommandAsync(npmCommandRunner: trm.ToolRunner, execOptions : trm.IExecOptions) : Promise<number> {
+    try{
+        var code: number = await npmCommandRunner.exec(execOptions);
+        tl.debug('Npm command succeeded with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        tl.debug(tl.loc('NpmFailed', err.message));
+        throw err;
+    }
+}
+
+function shouldUseDeprecatedTask() : boolean {
+    return tl.getVariable('USE_DEPRECATED_TASK_VERSION') != undefined && tl.getVariable('USE_DEPRECATED_TASK_VERSION').toLowerCase() === 'true';
+}
+
+function getNpmAuthHelperRunner(npmrcPath: string, tempUserNpmrcPath: string, includeDebugLogs: boolean): trm.ToolRunner {
+    var npmAuthHelper = tl.tool(path.join(__dirname, 'Npm/vsts-npm-auth/bin/vsts-npm-auth.exe'));
+    var verbosityString = includeDebugLogs ? 'Detailed' : 'Normal';
+    npmAuthHelper.line(`-NonInteractive -Verbosity ${verbosityString} -Config "${npmrcPath}" -TargetConfig "${tempUserNpmrcPath}"`);
+    return npmAuthHelper;
+}
+
+function getNpmConfigRunner(includeDebugLogs: boolean): trm.ToolRunner {
+    var npmConfig = tl.tool(tl.which('npm', true));
+    npmConfig.line('config list');
+    if(includeDebugLogs) {
+        npmConfig.arg('-l');
+    }
+    return npmConfig;
+}
+
+function getTempNpmrcPath() : string {
+    var tempNpmrcDir
+        = tl.getVariable('Agent.BuildDirectory')
+        || tl.getVariable('Agent.ReleaseDirectory')
+        || process.cwd();
+    tempNpmrcDir = path.join(tempNpmrcDir, 'npm');
+    tl.mkdirP(tempNpmrcDir);
+    var tempUserNpmrcPath: string = path.join(tempNpmrcDir, 'auth.' + tl.getVariable('build.buildId') + '.npmrc');
+    return tempUserNpmrcPath;
+}
+
+function cleanUpTempNpmrcPath(tempUserNpmrcPath: string) {
+    tl.debug('cleaning up...')
+    if(tl.exist(tempUserNpmrcPath)) {
+        tl.debug(`deleting temporary npmrc at '${tempUserNpmrcPath}'` );
+        tl.rmRF(tempUserNpmrcPath, /* continueOnError */ false);
+    }
+}
+
+function addBuildCredProviderEnv(env: EnvironmentDictionary) : EnvironmentDictionary {
+
+    let credProviderPath : string = path.join(__dirname, 'Npm/CredentialProvider');
 
-var cwd = tl.getPathInput('cwd', true, false);
-tl.mkdirP(cwd);
-tl.cd(cwd);
+    // get build access token
+    var accessToken : string = getSystemAccessToken();
```

Comment:
I know my own code isn't perfect here, but consider trying to standardize `let` vs `var`. 

Consider running tslint. If you're using VS Code there's an extension that does squigglies on tslint violations, and there are tslint config files in at least nuget-task-common.